### ISSUE TITLE
Enhancement Request: Speed up initial startup port collision checks

### ISSUE THREAD

**Issue by marblenix** [hashicorp/vagrant#11642]
**Is your feature request related to a problem? Please describe.**
When forwarding many ports, especially with multiple network interfaces, initial startup time can take a very long time. Roughly 1 second per port per interface.

**Describe the solution you'd like**
Faster startup time.

**Describe alternatives you've considered**
None

**Additional context**
running `vagrant up --debug` showed almost all the time being spent in the `handle_port_collisions` section of the code. A possible solution might be to decrease the timeout in this class: https://github.com/hashicorp/vagrant/blob/master/lib/vagrant/util/is_port_open.rb#L19

which can take a fraction for the timeout: https://ruby-doc.org/stdlib-2.7.1/libdoc/timeout/rdoc/Timeout.html

Possibly related: https://github.com/hashicorp/vagrant/issues/10729

**Comment by chrisroberts** [comment]
@marblenix Would you please provide a debug gist with timestamps enabled displaying the slowdown you are experiencing:

```
$ vagrant up --debug-timestamp
```

Thanks!

**Comment by marblenix** [comment]
@chrisroberts

```
everything before this took < 1 second


2020-05-26 12:45:44 [DEBUG] Waiting for process to exit. Remaining to timeout: 32000
2020-05-26 12:45:44 [DEBUG] Exit status: 0
2020-05-26 12:45:44 [ INFO] Calling IN action: #<Vagrant::Action::Builtin::HandleForwardedPortCollisions:0x0000000007705f60>
2020-05-26 12:45:44 [DEBUG] Attempting to acquire process-lock: fpcollision
2020-05-26 12:45:44 [DEBUG] Attempting to acquire process-lock: dotlock
2020-05-26 12:45:44 [ INFO] Acquired process lock: dotlock
2020-05-26 12:45:44 [ INFO] Released process lock: dotlock
2020-05-26 12:45:44 [ INFO] Acquired process lock: fpcollision
2020-05-26 12:45:44 [ INFO] Detecting any forwarded port collisions...
2020-05-26 12:45:44 [DEBUG] Extra in use: {}
2020-05-26 12:45:44 [DEBUG] Remap: {}
2020-05-26 12:45:44 [DEBUG] Repair: true
2020-05-26 12:45:44 [DEBUG] Testing port 80 on all IPv4 interfaces...
2020-05-26 12:45:44 [DEBUG] Testing ethernet_32770 with IP address [removed]
2020-05-26 12:45:46 [DEBUG] Testing ethernet_32774 with IP address [removed]
2020-05-26 12:45:48 [DEBUG] Testing ethernet_32769 with IP address [removed]
2020-05-26 12:45:50 [DEBUG] Testing loopback_0 with IP address 127.0.0.1
2020-05-26 12:45:52 [DEBUG] Port 80 will forward to the guest on the following interfaces: [["ethernet_32770", "[removed]"], ["ethernet_32774", "[removed]"], ["ethernet_32769", "[removed]"], ["loopback_0", "127.0.0.1"]]
2020-05-26 12:45:52 [DEBUG] Testing port 443 on all IPv4 interfaces...
2020-05-26 12:45:52 [DEBUG] Testing ethernet_32770 with IP address [removed]
2020-05-26 12:45:54 [DEBUG] Testing ethernet_32774 with IP address [removed]
2020-05-26 12:45:56 [DEBUG] Testing ethernet_32769 with IP address [removed]
2020-05-26 12:45:58 [DEBUG] Testing loopback_0 with IP address 127.0.0.1
2020-05-26 12:46:00 [DEBUG] Port 443 will forward to the guest on the following interfaces: [["ethernet_32770", "[removed]"], ["ethernet_32774", "[removed]"], ["ethernet_32769", "[removed]"], ["loopback_0", "127.0.0.1"]]
2020-05-26 12:46:24 [DEBUG] Testing port 8080 on all IPv4 interfaces...
2020-05-26 12:46:24 [DEBUG] Testing ethernet_32770 with IP address [removed]
2020-05-26 12:46:26 [DEBUG] Testing ethernet_32774 with IP address [removed]
2020-05-26 12:46:28 [DEBUG] Testing ethernet_32769 with IP address [removed]
2020-05-26 12:46:30 [DEBUG] Testing loopback_0 with IP address 127.0.0.1
2020-05-26 12:46:32 [DEBUG] Port 8080 will forward to the guest on the following interfaces: [["ethernet_32770", "[removed]"], ["ethernet_32774", "[removed]"], ["ethernet_32769", "[removed]"], ["loopback_0", "127.0.0.1"]]


... truncated to end of port testing


2020-05-26 12:58:44 [DEBUG] Waiting for process to exit. Remaining to timeout: 32000
2020-05-26 12:58:44 [DEBUG] Exit status: 0


... truncated to end of log


2020-05-26 12:59:57 [ INFO] Acquired process lock: dotlock
2020-05-26 12:59:57 [ INFO] Released process lock: dotlock
2020-05-26 12:59:57 [ INFO] Running hook: environment_unload
2020-05-26 12:59:57 [ INFO] Running action: environment_unload #<Vagrant::Action::Builder:0x0000000008fa0c78>
```

**Comment by marblenix** [comment]
Update: I've manually changed the file on disk to this and it's seems to have solved my issue. If you'd like I could make a pull request with the change.

```ruby
require "socket"

module Vagrant
  module Util
    # Contains the method {#is_port_open?} to check if a port is open
    # (listening) or closed (not in use). This method isn't completely
    # fool-proof, but it works enough of the time to be useful.
    module IsPortOpen
      # Checks if a port is open (listening) on a given host and port.
      # https://stackoverflow.com/a/3473208
      #
      # @param [String] host Hostname or IP address.
      # @param [Integer] port Port to check.
      # @return [Boolean] `true` if the port is open (listening), `false`
      #   otherwise.
      def is_port_open?(host, port)
        s = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM, 0)
        sa = Socket.sockaddr_in(port, host)

        begin
          s.connect_nonblock(sa)
        rescue Errno::EINPROGRESS
          if ::IO.select(nil, [s], nil, 0.1)
            begin
              s.connect_nonblock(sa)
            rescue Errno::EISCONN
              return true
            rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH
              return false
            end
          end
        end

        return false
      end
    end
  end
end
```

**Comment by chrisroberts** [comment]
Closing as fixed via #11810
